streamz/tests/test_kafka.py

```diff
@@ -0,0 +1,210 @@
+LAUNCH_KAFKA = os.environ.get('STREAMZ_LAUNCH_KAFKA', '') == 'true'
+TOPIC = 'test'
+ck = pytest.importorskip('confluent_kafka')
+
+
+def download_kafka(target):
+    r = requests.get('http://apache.mirror.globo.tech/kafka/1.0.0/'
+                     '%s.tgz' % KAFKA_FILE, stream=True)
+    with open(target, 'wb') as f:
+        for chunk in r.iter_content(2 ** 20):
+            f.write(chunk)
+    subprocess.check_call(['tar', 'xzf', KAFKA_FILE],
+                          cwd=os.path.dirname(target))
+
+
+def stop_docker(name='streamz-kafka', cid=None, let_fail=False):
+    """Stop docker container with given name tag
+
+    Parameters
+    ----------
+    name: str
+        name field which has been attached to the container we wish to remove
+    cid: str
+        container ID, if known
+    let_fail: bool
+        whether to raise an exception if the underlying commands return an
+        error.
+    """
+    try:
+        if cid is None:
+            print('Finding %s ...' % name)
+            cmd = shlex.split('docker ps -q --filter "name=%s"' % name)
+            cid = subprocess.check_output(cmd).strip().decode()
+        if cid:
+            print('Stopping %s ...' % cid)
+            subprocess.call(['docker', 'rm', '-f', cid])
+    except subprocess.CalledProcessError as e:
+        print(e)
+        if not let_fail:
+            raise
+
+
+def launch_kafka():
+    stop_docker(let_fail=True)
+    cmd = ("docker run -d -p 2181:2181 -p 9092:9092 --env "
+           "ADVERTISED_HOST=127.0.01 --env ADVERTISED_PORT=9092 "
+           "--name streamz-kafka spotify/kafka")
+    print(cmd)
+    cid = subprocess.check_output(shlex.split(cmd)).decode()[:-1]
+
+    def end():
+        if cid:
+            stop_docker(cid=cid)
+    atexit.register(end)
+
+    def predicate():
+        try:
+            out = subprocess.check_output(['docker', 'logs', cid],
+                                      stderr=subprocess.STDOUT)
+            return b'kafka entered RUNNING state' in out
+        except subprocess.CalledProcessError:
+            pass
+    wait_for(predicate, 10, period=0.1)
+    return cid
+
+
+def create_kafka_topic(cid, topic, partitions):
+    cmd = ("docker exec {cid} /opt/kafka_2.11-0.10.1.0/bin/kafka-topics.sh "
+           "--create --zookeeper localhost:2181 --replication-factor 1 "
+           "--partitions {parts} --topic {topic}".format(
+                cid=cid, parts=partitions, topic=topic))
+    print(cmd)
+    out = subprocess.check_output(shlex.split(cmd))
+    return b'Created topic' in out
+
+
+_kafka = [None]
+
+
+@contextmanager
+def kafka_service():
+    if _kafka[0] is None:
+        if LAUNCH_KAFKA:
+            launch_kafka()
+        else:
+            raise pytest.skip.Exception("Kafka not available. "
+                    "To launch kafka use `export STREAMZ_LAUNCH_KAFKA=true`")
+
+        producer = ck.Producer({'bootstrap.servers': 'localhost:9092'})
+        producer.produce('test-start-kafka', b'test')
```

Comment:
This ought to fail, as the topic does not exist. "test" is automatically created, and I included a `create_kafka_topic()` function in `test_kafka`, that's not actually called.

Reply:
How hard is it to destroy topics?  It would be nice to avoid reusing the same topic between tests.  Things seem to get a bit messy.

Reply:
confluent_kafka cannot create topics, I don't know if pykafka can - you need to call a docker exec as in the function I wrote, and a similar function with `--delete` would remove it. However, If we *only* run tests on docker, the issue goes away, we can use a new topic for every test.

Reply:
(i.e,. we kill the container after use, so all topics go away)